Add Python serialization to Pattern Matcher patterns

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109066
* #109156
* #109142
* #108917
* __->__ #108894
* #109663

Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

eellison committed:
Add Pretty Python Print to serialize patterns

[ghstack-poisoned]

eellison committed:
Update on "Add Pretty Python Print to serialize patterns"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff. Then I will include all inference patterns, then all training patterns.


Example Serialized Pattern:

```
tmp_0 = CallFunction(
    aten.view.default,
    CallFunction(aten.expand.default, KeywordArg("query"), Ignored()),
    Ignored(),
    _users=2,
)
tmp_1 = CallFunction(
    aten.view.default,
    CallFunction(
        aten.expand.default,
        CallFunction(aten.permute.default, KeywordArg("key"), Ignored()),
        Ignored(),
    ),
    Ignored(),
    _users=2,
)
tmp_2 = CallFunction(
    aten.div.Tensor,
    CallFunction(
        aten.view.default, CallFunction(aten.bmm.default, tmp_0, tmp_1), Ignored()
    ),
    KeywordArg("inv_scale"),
    _users=2,
)
tmp_3 = CallFunction(
    aten.exp.default,
    CallFunction(
        aten.sub.Tensor, tmp_2, CallFunction(aten.amax.default, tmp_2, Ignored(), True)
    ),
    _users=2,
)
tmp_4 = CallFunction(
    aten.div.Tensor,
    tmp_3,
    CallFunction(aten.sum.dim_IntList, tmp_3, Ignored(), True),
    _users=3,
)
tmp_5 = CallFunction(
    aten.view.default,
    CallFunction(aten.expand.default, tmp_4, Ignored()),
    Ignored(),
    _users=2,
)
tmp_6 = CallFunction(
    aten.view.default,
    CallFunction(aten.expand.default, KeywordArg("value"), Ignored()),
    Ignored(),
    _users=2,
)
tmp_7 = CallFunction(aten.view.default, KeywordArg("tangents_1"), Ignored(), _users=2)
tmp_8 = CallFunction(
    aten.mul.Tensor,
    CallFunction(
        aten.view.default,
        CallFunction(
            aten.bmm.default,
            tmp_7,
            CallFunction(aten.permute.default, tmp_6, Ignored()),
        ),
        Ignored(),
    ),
    tmp_4,
    _users=2,
)
tmp_9 = CallFunction(
    aten.view.default,
    CallFunction(
        aten.div.Tensor,
        CallFunction(
            aten.sub.Tensor,
            tmp_8,
            CallFunction(
                aten.mul.Tensor,
                tmp_4,
                CallFunction(aten.sum.dim_IntList, tmp_8, Ignored(), True),
            ),
        ),
        KeywordArg("inv_scale"),
    ),
    Ignored(),
    _users=2,
)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/fx_passes/serialized_attention_patterns/_sfdp_pattern_1.py:
Should we make this directory more general than "attention_patterns"?  I suspect there will be other cases where we want to use this too.  Perhaps just "serialized_patterns"?

jansel commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -552,6 +579,55 @@ def _match(self, node: torch.fx.Node, ctx: MatchContext):
+        return "\n".join(output)
+
+    def pretty_print(self, obj, level):
+        if isinstance(obj, _TargetArgsExpr):
```
Why not just memoize all patterns?

eellison replied:
The memoization is needed for patterns that are used multiple times. IMO if we memoized everything it would make the patterns less readable since all CallFunctions would get written out to a temporary.

Example Before and After.

Before:

```
tmp_0 = CallFunction(aten.div.Tensor,
  CallFunction(aten.view.default,
    CallFunction(aten.bmm.default,
      CallFunction(aten.view.default,
        CallFunction(aten.expand.default,
          KeywordArg('query'),
          Ignored()
        ),
        Ignored()
      ),
      CallFunction(aten.view.default,
        CallFunction(aten.expand.default,
          CallFunction(aten.permute.default,
            KeywordArg('key'),
            Ignored()
          ),
          Ignored()
        ),
        Ignored()
      )
    ),
    Ignored()
  ),
  KeywordArg('inv_scale'),
  _users=2
)
tmp_1 = CallFunction(aten.exp.default,
  CallFunction(aten.sub.Tensor,
    tmp_0,
    CallFunction(aten.amax.default,
      tmp_0,
      Ignored(),
      True
    )
  ),
  _users=2
)
_sfdp_pattern_1_inference = CallFunction(aten.view.default,
  CallFunction(aten.bmm.default,
    CallFunction(aten.view.default,
      CallFunction(aten.expand.default,
        CallFunction(aten.div.Tensor,
          tmp_1,
          CallFunction(aten.sum.dim_IntList,
            tmp_1,
            Ignored(),
            True
          )
        ),
        Ignored()
      ),
      Ignored()
    ),
    CallFunction(aten.view.default,
      CallFunction(aten.expand.default,
        KeywordArg('value'),
        Ignored()
      ),
      Ignored()
    )
  ),
  Ignored(),
  _users=0
)
```

After:

```
tmp_0 = CallFunction(aten.expand.default,
  KeywordArg('query'),
  Ignored()
)
tmp_1 = CallFunction(aten.view.default,
  tmp_0,
  Ignored()
)
tmp_2 = CallFunction(aten.permute.default,
  KeywordArg('key'),
  Ignored()
)
tmp_3 = CallFunction(aten.expand.default,
  tmp_2,
  Ignored()
)
tmp_4 = CallFunction(aten.view.default,
  tmp_3,
  Ignored()
)
tmp_5 = CallFunction(aten.bmm.default,
  tmp_1,
  tmp_4
)
tmp_6 = CallFunction(aten.view.default,
  tmp_5,
  Ignored()
)
tmp_7 = CallFunction(aten.div.Tensor,
  tmp_6,
  KeywordArg('inv_scale'),
  _users=2
)
tmp_8 = CallFunction(aten.amax.default,
  tmp_7,
  Ignored(),
  True
)
tmp_9 = CallFunction(aten.sub.Tensor,
  tmp_7,
  tmp_8
)
tmp_10 = CallFunction(aten.exp.default,
  tmp_9,
  _users=2
)
tmp_11 = CallFunction(aten.sum.dim_IntList,
  tmp_10,
  Ignored(),
  True
)
tmp_12 = CallFunction(aten.div.Tensor,
  tmp_10,
  tmp_11
)
tmp_13 = CallFunction(aten.expand.default,
  tmp_12,
  Ignored()
)
tmp_14 = CallFunction(aten.view.default,
  tmp_13,
  Ignored()
)
tmp_15 = CallFunction(aten.expand.default,
  KeywordArg('value'),
  Ignored()
)
tmp_16 = CallFunction(aten.view.default,
  tmp_15,
  Ignored()
)
tmp_17 = CallFunction(aten.bmm.default,
  tmp_14,
  tmp_16
)
_sfdp_pattern_1_inference = CallFunction(aten.view.default,
  tmp_17,
  Ignored(),
  _users=0
)
```

I don't have a strong preference here and happy to do either.

eellison replied:
I guess we could use more semantic names then `tmp` and it would be more readable

eellison replied:
switched to using semantic name, looks good now

jansel reviewed (dismissed):
Do we need a test to ensure the pattern file is up to date?

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jansel commented on test/inductor/test_pattern_matcher.py:
```diff
@@ -776,6 +785,63 @@ def fn2(inp, a, b):
+                    continue
+
+                self.assertEqual(
+                    pattern_pp, PatternPrettyPrinter.run(search_fn_pattern)
```
Add a `msg=` telling people the command to run to update the patterns.

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge -f "flakey failures, working later in stack"

eellison commented:
@pytorchbot revert -m "land race" -c landrace

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add Python serialization to Pattern Matcher patterns"


Adds a Python Pretty Printer to the pattern matcher that serializes patterns as python. Generating our fuse attention patterns was taking 4 seconds of compile time, which will only get worse as we add more variants (which I will do in the rest of this stack). To write out patterns, build pytorch, then run `gen_attention_patterns.py`.

Since there is a line limit for PRs  i'm only including the _sdpa_pattern1 in this first diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
